lectures/zksnark.tex

```diff
@@ -0,0 +1,812 @@
+            u_2 v_1 & u_2 v_2 & \cdots & u_2 v_n \\
+            \vdots & \vdots & \ddots & \vdots \\
+            u_m v_1 & u_m v_2 & \cdots & u_m v_n
+        \end{pmatrix}
+    \end{equation*}
+    With the following properties $\forall (c, u, v, w) \in \mathbb{F} \times \mathbb{F}^n \times \mathbb{F}^m \times \mathbb{F}^p$:
+    \begin{itemize}
+        \item Transpose: $(u \otimes v) = (v \otimes u)^{\textsf{T}}$
+        \item Distributivity: $u \otimes (v + w) = u \otimes v + u \otimes w$
+        \item Scalar Multiplication: $c(v \otimes u) = (cv) \otimes u = v \otimes (cu)$
+        \item Rank: the outer product $u \otimes v$ is a rank-1 matrix if $u$ and $v$ are non-zero
+        vectors
+    \end{itemize}
+\end{definition}
+
+\begin{example}
+    Let $u, v$ are vectors over the real number $\mathbb{R}$, where
+    \begin{equation*}
+        u = (1, 2, 3), \quad v = (2, 4, 3)
+    \end{equation*}
+    Then: 
+    \begin{equation*}
+        u \otimes v = \begin{pmatrix}
+            1 \cdot 2 & 1 \cdot 4 & 1 \cdot 3 \\
+            2 \cdot 2 & 2 \cdot 4 & 2 \cdot 3 \\
+            3 \cdot 2 & 3 \cdot 4 & 3 \cdot 3
+        \end{pmatrix} = \begin{pmatrix}
+            2 & 4 & 3 \\
+            4 & 8 & 6 \\
+            6 & 12 & 9
+        \end{pmatrix}
+    \end{equation*}
+    Additionally, as we can see the rows number 2 and 3 in the result matrix can be represented
+    as a linear combination of the first row, specifically by multiplying it by 2 and 3, 
+    respectively. The same property applies to the columns. This demonstrates the property of the
+    outer product, that the resulting matrix has a rank of 1.
+\end{example}
+
+Using the outer product we can express the constraint in another form.
+
+\begin{lemma}
+    Suppose there is a constraint $\langle a, w\rangle \times b, w\rangle = \langle c, w \rangle$ 
+    with coefficient vectors $a$, $b$, $c$ and witness vector $w$. Then it can be expressed in the 
+    form:
+    \[ w^T A w + c^T w = 0 \]
+    Where $A$ is the outer product of vectors $a$, $b$ (denoted as $a \otimes b$), consequently 
+    a \textbf{rank-1} matrix.
+\end{lemma}
+
+\textbf{Lemma proof.} Suppose there is a constraint $\langle a, w\rangle \times \langle b, w\rangle
+= \langle c, w\rangle$. Where vectors $a, b, c, w \in \mathbb{F}^n $. Let's expand the inner
+products: 
+\[ \sum_{i=1}^{n} a_i w_i \times \sum_{j=1}^{n} b_j w_j = \sum_{k=1}^{n} c_k w_k \]
+
+Combine the products into a double sum on the left side:
+\[ \sum_{i=1}^{n} \sum_{j=1}^{n} a_i b_j w_i w_j = w^T (a \otimes b) w = w^T A w \]
+
+Thus, the constraint can be written as:
+\[ w^T A w + c^T w = 0 \]
+
+So, the rank-1 means the rank of the coefficients matrix $A$ in one of the constraint formats.
+
+\subsection{Quadratic Arithmetic Program}
+
+While the Rank-1 Constraint System provides a powerful way to represent computations, it is not 
+succinct at all, since the number of constraints directly depends on the complexity of the problem
```

Comment:
"Directly depends" is a bit vague term. I believe it would be appropriate to say that it "depends linearly on the complexity of the problem being solved."